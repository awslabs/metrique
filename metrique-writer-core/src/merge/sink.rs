// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! Entry sink that aggregates entries before forwarding to a downstream sink.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use crate::EntrySink;
use super::{AggregatableEntry, AggregatedEntry};

/// A sink that aggregates entries with the same sample group before forwarding.
///
/// Entries are accumulated in memory and periodically flushed to the downstream sink.
pub struct AggregatingEntrySink<E: AggregatableEntry, S> {
    inner: Arc<AggregatingEntrySinkInner<E, S>>,
}

struct AggregatingEntrySinkInner<E: AggregatableEntry, S> {
    downstream: S,
    state: Mutex<AggregateState<E>>,
    config: AggregateConfig,
}

struct AggregateState<E: AggregatableEntry> {
    /// Aggregated entries keyed by their Key type
    aggregated: HashMap<E::Key, E::Aggregated>,
}

/// Configuration for aggregation behavior.
#[derive(Debug, Clone)]
pub struct AggregateConfig {
    /// Maximum number of entries to merge before flushing.
    pub max_entries: usize,
    /// Sample rate for emitting unmerged entries (0.0 = none, 1.0 = all).
    pub sample_rate: f64,
}

impl Default for AggregateConfig {
    fn default() -> Self {
        Self {
            max_entries: 100,
            sample_rate: 0.01,
        }
    }
}

impl<E: AggregatableEntry, S> AggregatingEntrySink<E, S>
where
    S: EntrySink<E::Aggregated>,
{
    /// Create a new aggregating sink with default configuration.
    pub fn new(downstream: S) -> Self {
        Self::with_config(downstream, AggregateConfig::default())
    }

    /// Create a new aggregating sink with custom configuration.
    pub fn with_config(downstream: S, config: AggregateConfig) -> Self {
        Self {
            inner: Arc::new(AggregatingEntrySinkInner {
                downstream,
                state: Mutex::new(AggregateState {
                    aggregated: HashMap::new(),
                }),
                config,
            }),
        }
    }

    /// Flush all accumulated merged entries to the downstream sink.
    pub fn flush_aggregated(&self) {
        let mut state = self.inner.state.lock().unwrap();
        for (_, merged) in state.aggregated.drain() {
            self.inner.downstream.append(merged);
        }
    }
}

impl<E: AggregatableEntry, S> EntrySink<E> for AggregatingEntrySink<E, S>
where
    S: EntrySink<E::Aggregated>,
{
    fn append(&self, entry: E) {
        let mut state = self.inner.state.lock().unwrap();
        
        // Extract key from entry
        let key = entry.key();

        // Get or create merged entry for this key
        let merged = state
            .aggregated
            .entry(key.clone())
            .or_insert_with(|| E::new_aggregated(key));

        merged.aggregate_into(&entry);

        // Check if we should flush
        if state.aggregated.len() >= self.inner.config.max_entries {
            drop(state);
            self.flush_aggregated();
        }
    }

    fn flush_async(&self) -> crate::sink::FlushWait {
        self.flush_aggregated();
        self.inner.downstream.flush_async()
    }
}
