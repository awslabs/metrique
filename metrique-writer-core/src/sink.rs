// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! Contains the [`EntrySink`] trait, which provides sinks into which metric entries
//! can be written. Unlike [`EntryIoStream`], these can be asynchronous.
//!
//! [`EntryIoStream`]: crate::stream::EntryIoStream

use std::{
    fmt::Debug,
    ops::{Deref, DerefMut},
    pin::Pin,
    sync::Arc,
};

use crate::{Entry, entry::BoxEntry};

/// Stores entries in an in-memory buffer until they can be written to the destination.
///
/// Implementations of this trait normally manage a queueing policy, then pass the [`Entry`]
/// to an [`EntryIoStream`] (in `metrique-writer`, there is `FlushImmediately` with a trivial queueing
/// policy, and `BackgroundQueue` which flushes entries via a queue).
///
/// [`EntryIoStream`]: crate::stream::EntryIoStream
pub trait EntrySink<E: Entry> {
    /// Append the `entry` to the in-memory buffer. Unless this is explicitly a test sink, the `append()` call must
    /// never block and must never panic. Test sinks are encouraged to immediately panic on invalid entries. Production
    /// sinks should emit a `tracing` event when invalid entries are found.
    ///
    /// If the in-memory buffer is bounded and full, the oldest entries should be dropped. More recent entries are more
    /// valuable for monitoring service health.
    fn append(&self, entry: E);

    /// Request the sink to flush its contents to some sort of persistent storage. The returned
    /// `FlushWait` can be used to tell when the sink is flushed.
    ///
    /// In synchronous code, you can use `pollster::block_on` or `futures::executor::block_on` to
    /// wait for this future to complete.
    fn flush_async(&self) -> FlushWait;

    /// Wrap `entry` in a smart pointer that will automatically append it to this sink when dropped.
    ///
    /// This will help enforce that an entry is always appended even if it's used across branching business logic. Note
    /// that Rust can't guarantee that the entry is dropped (e.g. `forget(entry)`).
    ///
    /// # Example
    /// ```
    /// # use metrique_writer::{Entry, sink::VecEntrySink, EntrySink};
    /// #[derive(Entry, PartialEq, Debug)]
    /// struct MyEntry {
    ///     counter: u64,
    /// }
    ///
    /// let sink = VecEntrySink::default();
    /// {
    ///     let mut entry = sink.append_on_drop(MyEntry { counter: 0 });
    ///     // do some business logic
    ///     entry.counter += 1;
    /// }
    /// assert_eq!(sink.drain(), &[MyEntry { counter: 1}]);
    /// ```
    fn append_on_drop(&self, entry: E) -> AppendOnDrop<E, Self>
    where
        Self: Sized + Clone,
    {
        AppendOnDrop::new(entry, self.clone())
    }

    /// See [`EntrySink::append_on_drop()`].
    fn append_on_drop_default(&self) -> AppendOnDrop<E, Self>
    where
        Self: Sized + Clone,
        E: Default,
    {
        self.append_on_drop(E::default())
    }
}

/// Provides a more generic interface than [`EntrySink`] but may come at the cost of dynamic dispatch and heap
/// allocation to store the in-memory buffer.
pub trait AnyEntrySink {
    /// Generic version of [`EntrySink::append()`] with the same contract.
    fn append_any(&self, entry: impl Entry + Send + 'static);

    /// Request the sink to flush its contents and wait until they are flushed.
    ///
    /// In synchronous code, you can use `pollster::block_on` or `futures::executor::block_on` to
    /// wait for this future to complete.
    fn flush_async(&self) -> FlushWait;

    /// Returns a [`BoxEntrySink`] that is a type-erased version of this entry sink
    fn boxed(self) -> BoxEntrySink
    where
        Self: Sized + Send + Sync + 'static,
    {
        BoxEntrySink::new(self)
    }
}

impl<T: AnyEntrySink, E: Entry + Send + 'static> EntrySink<E> for T {
    fn flush_async(&self) -> FlushWait {
        AnyEntrySink::flush_async(self)
    }

    fn append(&self, entry: E) {
        self.append_any(entry)
    }
}

/// A type-erased [`EntrySink`], that can sink a [`BoxEntry`] (which can contain
/// an arbitrary [`Entry`]).
#[derive(Clone)]
pub struct BoxEntrySink(Arc<Box<dyn EntrySink<BoxEntry> + Send + Sync + 'static>>);

impl Debug for BoxEntrySink {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("BoxEntrySink").finish()
    }
}

impl AnyEntrySink for BoxEntrySink {
    fn append_any(&self, entry: impl Entry + Send + 'static) {
        self.0.append(entry.boxed())
    }

    fn flush_async(&self) -> FlushWait {
        self.0.flush_async()
    }
}

impl BoxEntrySink {
    /// Create a new [BoxEntrySink]
    pub fn new(sink: impl EntrySink<BoxEntry> + Send + Sync + 'static) -> Self {
        Self(Arc::new(Box::new(sink)))
    }
}

/// This struct contains a future that can be used to wait for flushing to complete
#[must_use = "future does nothing unless polled"]
pub struct FlushWait(Pin<Box<dyn std::future::Future<Output = ()> + 'static>>);

impl Future for FlushWait {
    type Output = ();

    fn poll(
        mut self: Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Self::Output> {
        self.0.as_mut().poll(cx)
    }
}

impl FlushWait {
    /// Return a FlushWait that is ready immediately
    pub fn ready() -> Self {
        // VecEntrySink is synchronous, poll_fn is zero_sized unlike Ready<()>
        Self(Box::pin(std::future::poll_fn(|_| {
            std::task::Poll::Ready(())
        })))
    }

    /// Create a FlushWait that returns when a future is ready
    pub fn from_future(f: impl std::future::Future<Output = ()> + 'static) -> Self {
        Self(Box::pin(f))
    }
}

/// Smart pointer that will append the wrapped entry to a sink when dropped.
#[derive(Debug, Clone)]
pub struct AppendOnDrop<E: Entry, Q: EntrySink<E>> {
    entry: Option<E>,
    sink: Q,
}

impl<E: Entry, Q: EntrySink<E>> AppendOnDrop<E, Q> {
    pub(crate) fn new(entry: E, sink: Q) -> Self {
        Self {
            entry: Some(entry),
            sink,
        }
    }
}

impl<E: Entry, Q: EntrySink<E>> Drop for AppendOnDrop<E, Q> {
    fn drop(&mut self) {
        if let Some(entry) = self.entry.take() {
            self.sink.append(entry)
        }
    }
}

impl<E: Entry, Q: EntrySink<E>> AppendOnDrop<E, Q> {
    /// Take and return the entry out of this [AppendOnDrop], without
    /// appending it to the sink
    pub fn into_entry(mut self) -> E {
        self.entry.take().unwrap()
    }

    /// Drop the entry, but don't append it to the sink.
    pub fn forget(mut self) {
        self.entry = None;
    }
}

impl<E: Entry, Q: EntrySink<E>> Deref for AppendOnDrop<E, Q> {
    type Target = E;

    fn deref(&self) -> &Self::Target {
        self.entry.as_ref().unwrap()
    }
}

impl<E: Entry, Q: EntrySink<E>> DerefMut for AppendOnDrop<E, Q> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.entry.as_mut().unwrap()
    }
}
