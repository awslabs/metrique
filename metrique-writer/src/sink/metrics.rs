// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/// Defines callbacks for recording metrics
pub trait MetricRecorder: Send + Sync {
    /// Records a histogram entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32);
    /// Increments a histogram entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64);
    /// Sets a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64);
    /// Increments a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64);
    /// Decrements a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64);
}

/// Implements MetricRecorder for a global metrics-rs 0.24 recorder
#[cfg(feature = "metrics-rs-024")]
#[derive(Debug, Copy, Clone)]
pub(crate) struct GlobalMetricsRs024Bridge;

#[cfg(feature = "metrics-rs-024")]
impl MetricRecorder for GlobalMetricsRs024Bridge {
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32) {
        metrics_024::histogram!(metric, "sink" => sink.to_owned()).record(value);
    }

    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64) {
        metrics_024::counter!(metric, "sink" => sink.to_owned()).increment(value);
    }

    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics_024::gauge!(metric, "sink" => sink.to_owned()).set(value);
    }

    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics_024::gauge!(metric, "sink" => sink.to_owned()).increment(value);
    }

    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics_024::gauge!(metric, "sink" => sink.to_owned()).decrement(value);
    }
}

/// Implements MetricRecorder for a local metrics-rs 0.24 recorder
#[cfg(feature = "metrics-rs-024")]
#[derive(Debug, Copy, Clone)]
pub(crate) struct LocalMetricsRs024Bridge<R>(pub(crate) R);

#[cfg(feature = "metrics-rs-024")]
impl<R: metrics_024::Recorder + Send + Sync> MetricRecorder for LocalMetricsRs024Bridge<R> {
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32) {
        self.0
            .register_histogram(
                &metrics_024::Key::from_parts(
                    metric,
                    vec![metrics_024::Label::new("sink", sink.to_owned())],
                ),
                &metrics_024::Metadata::new(
                    module_path!(),
                    metrics_024::Level::INFO,
                    Some(module_path!()),
                ),
            )
            .record(value);
    }

    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64) {
        self.0
            .register_counter(
                &metrics_024::Key::from_parts(
                    metric,
                    vec![metrics_024::Label::new("sink", sink.to_owned())],
                ),
                &metrics_024::Metadata::new(
                    module_path!(),
                    metrics_024::Level::INFO,
                    Some(module_path!()),
                ),
            )
            .increment(value);
    }

    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics_024::Key::from_parts(
                    metric,
                    vec![metrics_024::Label::new("sink", sink.to_owned())],
                ),
                &metrics_024::Metadata::new(
                    module_path!(),
                    metrics_024::Level::INFO,
                    Some(module_path!()),
                ),
            )
            .set(value);
    }

    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics_024::Key::from_parts(
                    metric,
                    vec![metrics_024::Label::new("sink", sink.to_owned())],
                ),
                &metrics_024::Metadata::new(
                    module_path!(),
                    metrics_024::Level::INFO,
                    Some(module_path!()),
                ),
            )
            .increment(value);
    }

    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics_024::Key::from_parts(
                    metric,
                    vec![metrics_024::Label::new("sink", sink.to_owned())],
                ),
                &metrics_024::Metadata::new(
                    module_path!(),
                    metrics_024::Level::INFO,
                    Some(module_path!()),
                ),
            )
            .decrement(value);
    }
}

pub(crate) trait GlobalRecorderVersion {
    fn recorder() -> impl MetricRecorder + 'static;
    fn describe(metrics: &[DescribedMetric]);
}

#[cfg(feature = "metrics-rs-024")]
impl GlobalRecorderVersion for dyn metrics_024::Recorder {
    fn describe(metrics: &[DescribedMetric]) {
        for metric in metrics {
            let unit = match metric.unit {
                MetricsRsUnit::Count => metrics_024::Unit::Count,
                MetricsRsUnit::Percent => metrics_024::Unit::Percent,
                MetricsRsUnit::Millisecond => metrics_024::Unit::Milliseconds,
            };
            match metric.r#type {
                MetricsRsType::Counter => {
                    metrics_024::describe_counter!(metric.name, unit, metric.description)
                }
                MetricsRsType::Gauge => {
                    metrics_024::describe_gauge!(metric.name, unit, metric.description)
                }
                MetricsRsType::Histogram => {
                    metrics_024::describe_histogram!(metric.name, unit, metric.description)
                }
            }
        }
    }

    fn recorder() -> impl MetricRecorder {
        GlobalMetricsRs024Bridge
    }
}

/// Describes a metrics.rs unit in a non-exhaustive fashion
#[non_exhaustive]
#[derive(Copy, Clone, Debug)]
pub enum MetricsRsUnit {
    Percent,
    Count,
    Millisecond,
}

/// Describes a metrics.rs metric type in a non-exhaustive fashion
#[non_exhaustive]
#[derive(Copy, Clone, Debug)]
pub enum MetricsRsType {
    Gauge,
    Counter,
    Histogram,
}

#[derive(Copy, Clone, Debug)]
#[non_exhaustive]
pub struct DescribedMetric {
    pub name: &'static str,
    pub unit: MetricsRsUnit,
    pub r#type: MetricsRsType,
    pub description: &'static str,
}

pub trait LocalRecorderVersion<R> {
    fn recorder(recorder: R) -> impl MetricRecorder + 'static;
}

#[cfg(feature = "metrics-rs-024")]
impl<R> LocalRecorderVersion<R> for dyn metrics_024::Recorder
where
    R: metrics_024::Recorder + Send + Sync + 'static,
{
    fn recorder(recorder: R) -> impl MetricRecorder + 'static {
        LocalMetricsRs024Bridge(recorder)
    }
}
