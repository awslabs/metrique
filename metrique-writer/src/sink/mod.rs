// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! Contains various utilities for working with [EntrySink]

use std::sync::{Arc, Mutex};

use crate::Entry;

#[cfg(feature = "background_queue")]
mod background;
mod immediate_flush;
mod metrics;

#[cfg(feature = "background_queue")]
pub use background::{BACKGROUND_QUEUE_METRICS, describe_sink_metrics};
#[cfg(feature = "background_queue")]
pub use background::{BackgroundQueue, BackgroundQueueBuilder, BackgroundQueueJoinHandle};
pub use immediate_flush::{
    AnyFlushImmediately, FlushImmediately, FlushImmediatelyBuilder,
    describe_immediate_flush_metrics,
};
pub use metrique_writer_core::sink::{AnyEntrySink, AppendOnDrop, FlushWait};
use metrique_writer_core::{BoxEntrySink, EntryIoStream, EntrySink};
pub use metrique_writer_core::{
    global::AttachGlobalEntrySink, global::AttachHandle, global_entry_sink,
};

/// Extension trait for `AttachGlobalEntrySink`, containing functions that use
/// types that are not present in [`metrique_writer_core`].
pub trait AttachGlobalEntrySinkExt: AttachGlobalEntrySink {
    /// Attach the given output stream to a default [`BackgroundQueue`] and then to this
    /// global queue reference.
    ///
    /// # Panics
    /// Panics if a queue is already attached.
    fn attach_to_stream(output: impl EntryIoStream + Send + 'static) -> AttachHandle {
        Self::attach(BackgroundQueue::new(output))
    }
}

impl<Q: AttachGlobalEntrySink + ?Sized> AttachGlobalEntrySinkExt for Q {}

/// In-memory sink backed by a [`Vec`] designed for testing.
///
/// Cloning will provide another reference to the same underlying sink.
///
/// # Example
/// ```
/// # use metrique_writer::{Entry, EntrySink, sink::VecEntrySink};
/// #[derive(Entry, PartialEq, Debug)]
/// struct MyEntry { counter: u64 }
///
/// let sink = VecEntrySink::default();
/// sink.append(MyEntry { counter: 21 });
/// sink.append(MyEntry { counter: 42 });
/// assert_eq!(sink.drain(), &[MyEntry { counter: 21 }, MyEntry { counter: 42 }]);
/// ```
#[derive(Debug)]
pub struct VecEntrySink<E>(Arc<Mutex<Vec<E>>>);

impl<E> Default for VecEntrySink<E> {
    fn default() -> Self {
        Self(Default::default())
    }
}

impl<E> Clone for VecEntrySink<E> {
    fn clone(&self) -> Self {
        Self(Arc::clone(&self.0))
    }
}

impl<E: Entry> EntrySink<E> for VecEntrySink<E> {
    fn append(&self, entry: E) {
        self.0.lock().unwrap().push(entry);
    }

    fn flush_async(&self) -> FlushWait {
        FlushWait::ready()
    }
}

impl<E> VecEntrySink<E> {
    /// Create a new, empty [VecEntrySink]
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a new [`VecEntrySink`] using initial capacity for entries, to avoid
    /// unnecessary reallocations.
    ///
    /// The between this function and [`VecEntrySink::new`] is purely performance,
    /// in both cases, the [`VecEntrySink`] will resize itself if needed to hold
    /// a number of entries limited only by available memory.
    pub fn with_capacity(capacity: usize) -> Self {
        Self(Arc::new(Mutex::new(Vec::with_capacity(capacity))))
    }

    /// Drains all currently appended entries from the sink and returns them as an owned [`Vec`].
    ///
    /// The sink can still be used afterwards.
    pub fn drain(&self) -> Vec<E> {
        let mut entries = self.0.lock().unwrap();
        let empty = Vec::with_capacity(entries.capacity());
        std::mem::replace(&mut entries, empty)
    }

    /// Returns true if this [`VecEntrySink`] contains an entry which evaluates the predicate to true.
    pub fn contains_entry<F>(&self, predicate: F) -> bool
    where
        F: FnMut(&E) -> bool,
    {
        let entries = self.0.lock().unwrap();
        entries.iter().any(predicate)
    }
}

/// An [EntrySink] that drops all entries.
///
/// Useful for testing, or when you want to ignore entries.
#[derive(Copy, Clone, Default)]
#[non_exhaustive]
pub struct DevNullSink;

impl DevNullSink {
    /// Return a new [`DevNullSink`]
    pub const fn new() -> Self {
        DevNullSink
    }

    /// Return a new [`DevNullSink`] as a [`BoxEntrySink`]
    pub fn boxed() -> BoxEntrySink {
        Self::new().boxed()
    }
}

impl AnyEntrySink for DevNullSink {
    fn append_any(&self, _entry: impl Entry + Send + 'static) {}

    fn flush_async(&self) -> FlushWait {
        FlushWait::ready()
    }
}

#[cfg(test)]
mod tests {
    use std::time::SystemTime;

    use super::*;
    struct TestEntry {
        timestamp: SystemTime,
        counter: u32,
        status: String,
    }

    impl Entry for TestEntry {
        fn write<'a>(&'a self, writer: &mut impl crate::EntryWriter<'a>) {
            writer.timestamp(self.timestamp);
            writer.value("counter", &self.counter);
            writer.value("status", &self.status);
        }
    }

    #[test]
    fn vec_entry_sink_create_update_drain() {
        let sink = VecEntrySink::<TestEntry>::new();

        sink.append(TestEntry {
            timestamp: SystemTime::now(),
            counter: 1,
            status: "OK".into(),
        });
        sink.append(TestEntry {
            timestamp: SystemTime::now(),
            counter: 2,
            status: "OK".into(),
        });
        sink.append(TestEntry {
            timestamp: SystemTime::now(),
            counter: 0,
            status: "ERR".into(),
        });

        assert!(sink.contains_entry(|e| e.status == "ERR"));
        assert!(sink.contains_entry(|e| e.status == "OK" && e.counter != 0));

        let entries = sink
            .drain()
            .into_iter()
            .map(|e| (e.status, e.counter))
            .collect::<Vec<_>>();

        assert_eq!(
            entries,
            &[("OK".into(), 1), ("OK".into(), 2), ("ERR".into(), 0),]
        );

        assert!(!sink.contains_entry(|e| e.status == "ERR"));
        assert!(!sink.contains_entry(|e| e.status == "OK" && e.counter != 0));
        assert!(!sink.contains_entry(|_| true));
    }

    #[test]
    fn test_null_entry_sink() {
        let sink = DevNullSink::new();
        sink.append(TestEntry {
            timestamp: SystemTime::now(),
            counter: 1,
            status: "OK".into(),
        });
        futures::executor::block_on(EntrySink::<TestEntry>::flush_async(&sink));
    }
}
