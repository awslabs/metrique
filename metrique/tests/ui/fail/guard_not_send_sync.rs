use metrique_writer::sink::global_entry_sink;
use metrique_writer::test_util::{test_entry_sink, TestEntrySink};
global_entry_sink! { TestSink }

fn assert_sync<S: Sync>(_s: &S) {}
fn assert_send<S: Send>(_s: &S) {}

// this is not a test of the macro, but we already have a cfail test
fn main() {
    let TestEntrySink { inspector: _, sink } = test_entry_sink();
    let _guard = TestSink::set_test_sink(sink);
    assert_sync(&_guard);
    assert_send(&_guard);
}

