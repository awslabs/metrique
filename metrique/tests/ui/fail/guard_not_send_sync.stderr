error[E0277]: `*const ()` cannot be shared between threads safely
  --> tests/ui/fail/guard_not_send_sync.rs:12:17
   |
12 |     assert_sync(&_guard);
   |     ----------- ^^^^^^^ `*const ()` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `ThreadLocalTestSinkGuard`, the trait `Sync` is not implemented for `*const ()`
note: required because it appears within the type `PhantomData<*const ()>`
  --> $RUST/core/src/marker.rs
note: required because it appears within the type `ThreadLocalTestSinkGuard`
  --> $WORKSPACE/metrique-writer-core/src/global.rs
   |
   | pub struct ThreadLocalTestSinkGuard {
   |            ^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `assert_sync`
  --> tests/ui/fail/guard_not_send_sync.rs:5:19
   |
5  | fn assert_sync<S: Sync>(_s: &S) {}
   |                   ^^^^ required by this bound in `assert_sync`

error[E0277]: `*const ()` cannot be sent between threads safely
  --> tests/ui/fail/guard_not_send_sync.rs:13:17
   |
13 |     assert_send(&_guard);
   |     ----------- ^^^^^^^ `*const ()` cannot be sent between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `ThreadLocalTestSinkGuard`, the trait `Send` is not implemented for `*const ()`
note: required because it appears within the type `PhantomData<*const ()>`
  --> $RUST/core/src/marker.rs
note: required because it appears within the type `ThreadLocalTestSinkGuard`
  --> $WORKSPACE/metrique-writer-core/src/global.rs
   |
   | pub struct ThreadLocalTestSinkGuard {
   |            ^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `assert_send`
  --> tests/ui/fail/guard_not_send_sync.rs:6:19
   |
6  | fn assert_send<S: Send>(_s: &S) {}
   |                   ^^^^ required by this bound in `assert_send`
